### PR TITLE
[Fortran] Visit all the symbols and body in the Module and fix other bugs

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**

### CODE REVIEW

**Comment by certik** [comment]
Fixes like these will work most of the time, until they don't -- how will this handle if there is already a variable `v__t_1`?

**Comment by certik** [comment]
I will review this carefully later, but this approach is good, I think that's the way to do it.

**Comment by certik** [comment]
@Shaikh-Ubaid this is an example of a pass that we might want to run in the `--backend=fortran`, where we need GFortran to be able to compile it. But we don't want to run it by default, since we want any ASR to be able to print without additional passes.

**Comment by Shaikh-Ubaid** [comment]
I think we also need to handle `@` in functions from modules, for example:

```console
$ cat examples/expr2.f90 
program expr2
implicit none

print *, repeat("abc", 3)

end program
$ cat pass_27_insert_deallocate.f90 
! Fortran code after applying the pass: insert_deallocate
...
program expr2
implicit none
print *, repeat@repeati32("abc", 3)

end program expr2
```

**Comment by Thirumalai-Shaktivel** [comment]
With the current changes, I get:

Repeat
--
```fortran
program expr2
implicit none
print *, repeat("abc", 3)
end program
```
`--show-fortran --dump-all-passes-fortran`
```fortran
! Fortran code after applying the pass: insert_deallocate
module lfortran_intrinsic_iso_fortran_env
implicit none
integer(4), dimension(1), parameter :: character_kinds = [1]
integer(4), parameter :: error_unit = 0
integer(4), parameter :: input_unit = 5
integer(4), parameter :: int16 = 2
integer(4), parameter :: int32 = 4
integer(4), parameter :: int64 = 8
integer(4), parameter :: int8 = 1
integer(4), dimension(2), parameter :: integer_kinds = [4, 8]
integer(4), dimension(1), parameter :: logical_kinds = [4]
integer(4), parameter :: output_unit = 6
integer(4), parameter :: real128 = -1
integer(4), parameter :: real32 = 4
integer(4), parameter :: real64 = 8
integer(4), dimension(2), parameter :: real_kinds = [4, 8]
end module lfortran_intrinsic_iso_fortran_env

module lfortran_intrinsic_builtin
implicit none
end module lfortran_intrinsic_builtin

module lfortran_intrinsic_string
use lfortran_intrinsic_iso_fortran_env, only: int64
implicit none
interface repeat
    module procedure repeati32
end interface repeat

contains

elemental function len_repeati32(n) result(r)
    integer(4), intent(in) :: n
    integer(4) :: r
    r = n
end function len_repeati32
function repeati32(s, n) result(r)
    integer(4) :: i
    integer(4) :: i1
    integer(4), intent(in) :: n
    character(len=len_repeati32(n), kind=1) :: r
    character(len=1, kind=1), intent(in) :: s
    i1 = 1
    i = 1 - 1
    do while (i + 1 <= n)
        i = i + 1
        r(i:i) = s(i1:i1)
    end do
end function repeati32
end module lfortran_intrinsic_string

program expr2
use lfortran_intrinsic_string, only: len_repeati32
use lfortran_intrinsic_string, only: repeat
use lfortran_intrinsic_string, only: repeati32
implicit none
print *, repeat("abc", 3)
end program expr2
```
MaxVal
--
```fortran
program expr2
implicit none
integer :: x(3)
x = 23
end program
```
`--show-fortran --dump-all-passes-fortran --apply-fortran-mangling`
```fortran
! Fortran code after applying the pass: insert_deallocate
program expr2
implicit none
integer(4) :: v__1_t
integer(4) :: v__1_v
integer(4), dimension(3) :: x
v__1_v = lbound(x, 1)
v__1_t = lbound(x, 1) - 1
do while (v__1_t + 1 <= ubound(x, 1))
    v__1_t = v__1_t + 1
    x(v__1_t) = 23
    v__1_v = v__1_v + 1
end do
end program expr2
```
Both outputs get compiled with gfortran and print the expected output

This PR is ready for review; if there is anything else to be implemented, do let me know.

**Review by certik — APPROVED** [review]
As long as this mangling is not applied by default, which I think it is not, then this is fine. Thanks!

**Comment by certik** [comment]
Is the failure caused by this PR? It also needs to be rebased on top of the latest master.

**Comment by Thirumalai-Shaktivel** [comment]
The latest commit 8b40073d31e42526f09e6c0e86012e6471139e72 & 200966f82f985706ac71c11ce7acc56a9c982df5 seems to work for the Fortran backend, @certik What do you think?

**Comment by certik** [comment]
Excellent. I think it looks good in general. Go ahead and finish the PR and I give it a final review.

**Comment by Thirumalai-Shaktivel** [comment]
We can use `--committer-date-is-author-date` to not change the git commit date; that's a good one.

**Comment by Thirumalai-Shaktivel** [comment]
Ready for review!

File: integration_tests/CMakeLists.txt [hunk 1]
```diff
@@ -206,7 +206,7 @@ RUN(NAME include_03 LABELS gfortran llvm wasm c INCLUDE_PATH include_03 fortran)
 RUN(NAME cond_01 LABELS gfortran llvm cpp x86 wasm c fortran)
 RUN(NAME cond_02 LABELS gfortran llvm wasm c fortran)
 
-RUN(NAME expr_01 FAIL LABELS gfortran llvm cpp x86 wasm c fortran)
```
Comment (certik) [inline integration_tests/CMakeLists.txt:209]: Why is `fortran` removed?

**Comment by certik** [comment]
It seems this PR is implementing several unrelated fixes:

* Improvements to the Fortran backend (however no new tests are added, and old tests are removed)
* Different way of calling gfortran
* changes to the frontend's symbol table visitor
* improvements to the unique symbol pass to handle underscores

Can you split these into separate PRs? Then for each PR we'll make sure tests are added, etc. and we'll merge those that we can merge.

File: integration_tests/CMakeLists.txt (hunk 1 above)
Comment (Thirumalai-Shaktivel) [reply integration_tests/CMakeLists.txt:209]: When we link with the clang, gfortran doesn't generate the stack trace for the failure(It hangs), so I remove it.

**Comment by certik** [comment]
Thanks for splitting it. Now let's add tests for this.

**Comment by Thirumalai-Shaktivel** [comment]
Consider the test https://github.com/lfortran/lfortran/blob/main/integration_tests/complex_06.f90, which is already enabled to test the Fortran backend. It imports the `lfortran_*.f90` modules inside the program.
All the changes in this PR were introduced because of compiling the `lfortran_*.f90` module. So, I think everything is already tested, do you want me to enable other tests as well?

cc @certik

**Comment by certik** [comment]
@Thirumalai-Shaktivel go ahead and rebase this and enable Fortran tests that now work.

**Review by certik — APPROVED** [review]
I think that this looks good to merge, thanks!

**Pull request merged**